clust/utils.q

```diff
@@ -0,0 +1,1323 @@
+  validClusts:select from clusts[1]where valid;
+  $[dgram;
+    clust.i.dgramIdx last[clusts]0;
+    enlist @[;;:;]/[count[data 0]#0N;validClusts`points;til count validClusts]
+   ]
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Update dendrogram for CAW with final cluster of all the points
+// @param tab {table} Cluster table
+// @param linkMatrix {float[][]} Linkage matrix
+// @return {float[][]} Updated linkage matrix
+clust.i.updDgram:{[tab;linkMatrix]
+  linkMatrix,:value exec first clust,first nnIdx,first nnDist,count repPt 
+    from tab where nnDist=min nnDist;
+  linkMatrix
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Predict clusters using hierarchical or CURE config
+// @param name {symbol} Namespace to use - `hc or `cure
+// @param data {float[][]} Each column of the data is an individual datapoint
+// @param config {dictionary} Output of .ml.clust.(cutK/cutDist)
+// @return {long[]} Predicted clusters
+clust.i.hCCpred:{[name;data;config]
+  data:clust.i.floatConversion[data];
+  // Check correct namespace and clusters given
+  if[not name in`hc`cure;'"Incorrect namespace - please use `hc or `cure"];
+  if[not`clust in key config;
+    '"Clusters must be contained within config - please run .ml.clust.",
+    $[name~`hc;"hc";"cure"],".(cutK/cutDist)"];
+  // Add namespace and linkage to config dictionary for cure
+  if[name~`cure;config[`modelInfo;`inputs],:`name`lf!(name;`single)];
+  // Recalculate representative point for training clusters in asc
+  // order to ensure correct labels
+  clusts:group config`clust; 
+  clustKey:asc key clusts;
+  repPts:clust.i.getRep[config]each clusts clustKey;
+  // Training indices
+  idxs:til each numPt:count each repPts[;0];
+  // Return closest clusters to testing points
+  clust.i.predClosest[data;config;repPts;numPt;idxs]each til count data 0
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Recalculate representative points from training clusters
+// @param config {dictionary} Output of .ml.clust.(cutK/cutDist)
+// @param idxs {long[][]} Training data indices
+// @return {float[][]} Training data points
+clust.i.getRep:{[config;idxs]
+  config:config[`modelInfo];
+  $[config[`inputs;`name]~`cure;
+      flip(clust.i.cureRep . config[`inputs;`df`n`c])::;
+    config[`inputs;`lf]in`ward`centroid;
+      enlist each avg each;
+    ]config[`data][;idxs]
+  }
+
+// @private
+// @kind function
+// @category clustUtility
+// @desc Predict new cluster for given data point
+// @param data {float[][]} Each column of the data is an individual datapoint
+// @param config {dictionary} Output of .ml.clust.(cutK/cutDist)
+// @param repPt {float[][]} Representative points in matrix format
+// @param numPts {long} Number of points in training clusters
+// @param clustIdx {long[][]} Training data indices
+// @param ptIdx {long[][]} Index of current data point
+// @return {long[]} List of predicted clusters
```

Comment:
The return of this is the index of a single cluster, the nearest one, so should be defined as `index of the nearest cluster to the given point` (or something similar)